File: node/src/providers/provider-error.ts
export type ProviderErrorKind =
  | "transport"
  | "auth"
  | "rate_limit"
  | "context_overflow"
  | "invalid_request"
  | "modality"
  | "model_unavailable"
  | "protocol"
  | "unknown"

export interface ProviderErrorOptions {
  provider: string
  kind: ProviderErrorKind
  retryable: boolean
  message: string
  httpStatus?: number
  providerCode?: string
  cause?: unknown
}

/** Stable provider-failure contract. Only the named scalar fields cross the host ABI. */
export class ProviderError extends Error {
  readonly provider: string
  readonly kind: ProviderErrorKind
  readonly retryable: boolean
  readonly httpStatus?: number
  readonly providerCode?: string
  override readonly cause: unknown

  constructor(options: ProviderErrorOptions) {
    super(options.message, options.cause === undefined ? undefined : { cause: options.cause })
    this.name = "ProviderError"
    this.provider = options.provider
    this.kind = options.kind
    this.retryable = options.retryable
    this.httpStatus = options.httpStatus
    this.providerCode = options.providerCode
    this.cause = options.cause
  }
}

const CONTEXT_OVERFLOW_CODES = new Set([
  "context_length_exceeded",
  "prompt_too_long",
])
const NETWORK_CODES = new Set([
  "ECONNABORTED",
  "ECONNREFUSED",
  "ECONNRESET",
  "EHOSTUNREACH",
  "ENETUNREACH",
  "ETIMEDOUT",
])

function object(value: unknown): Record<string, unknown> | undefined {
  return value && typeof value === "object" ? value as Record<string, unknown> : undefined
}

function scalarString(value: unknown): string | undefined {
  return typeof value === "string" && value.length > 0 ? value : undefined
}

function scalarStatus(value: unknown): number | undefined {
  return typeof value === "number" && Number.isInteger(value) && value >= 100 && value <= 599
    ? value
    : undefined
}

function errorStatus(error: unknown): number | undefined {
  const outer = object(error)
  const response = object(outer?.response)
  return scalarStatus(outer?.httpStatus)
    ?? scalarStatus(outer?.status)
    ?? scalarStatus(response?.status)
}

function errorCode(error: unknown): string | undefined {
  const outer = object(error)
  const nested = object(outer?.error)
  const nestedError = object(nested?.error)
  const body = object(outer?.body)
  const bodyError = object(body?.error)
  return scalarString(outer?.providerCode)
    ?? scalarString(outer?.code)
    ?? scalarString(outer?.error_code)
    ?? scalarString(nested?.code)
    ?? scalarString(nested?.error_code)
    ?? scalarString(nestedError?.code)
    ?? scalarString(nestedError?.error_code)
    ?? scalarString(nestedError?.type)
    ?? scalarString(bodyError?.code)
    ?? scalarString(bodyError?.error_code)
}

function errorMessage(error: unknown): string {
  if (error instanceof Error && error.message) return error.message
  return typeof error === "string" ? error : "Provider request failed"
}

function classifyKind(error: unknown, status: number | undefined, code: string | undefined): ProviderErrorKind {
  if (code && CONTEXT_OVERFLOW_CODES.has(code.toLowerCase())) return "context_overflow"

  const name = error instanceof Error ? error.name : scalarString(object(error)?.name)
  if (name === "UnsupportedModalityError") return "modality"
  if (name === "ProtocolResponseError") return "protocol"
  if (name === "ContentValidationError" || name === "ProviderReplayValidationError") {
    return "invalid_request"
  }
  if (name === "APIConnectionError" || name === "APIConnectionTimeoutError") return "transport"

  if (status === 401 || status === 403) return "auth"
  if (status === 429) return "rate_limit"
  if (code === "model_not_found" || status === 404 || (status !== undefined && status >= 500)) {
    return "model_unavailable"
  }
  if (status === 408 || status === 409) return "transport"
  if (status === 400 || status === 422) return "invalid_request"

  const normalizedCode = code?.toUpperCase()
  if (normalizedCode && NETWORK_CODES.has(normalizedCode)) return "transport"
  if (error instanceof TypeError && status === undefined) return "transport"
  return "unknown"
}

function retryable(kind: ProviderErrorKind, status: number | undefined): boolean {
  if (kind === "transport" || kind === "rate_limit" || kind === "model_unavailable") return true
  if (kind === "unknown" && status !== undefined) return status >= 500 && status !== 501
  return false
}

export function classifyProviderError(provider: string, error: unknown): ProviderError {
  if (error instanceof ProviderError) return error
  const httpStatus = errorStatus(error)
  const providerCode = errorCode(error)
  const kind = classifyKind(error, httpStatus, providerCode)
  return new ProviderError({
    provider,
    kind,
    retryable: retryable(kind, httpStatus),
    message: errorMessage(error),
    ...(httpStatus !== undefined ? { httpStatus } : {}),
    ...(providerCode !== undefined ? { providerCode } : {}),
    cause: error,
  })
}

export function circuitOpenError(provider: string): ProviderError {
  return new ProviderError({
    provider,
    kind: "model_unavailable",
    retryable: true,
    providerCode: "circuit_open",
    message: "Circuit breaker open",
  })
}

/** Safe scalar projection for runner → canonical-host events. */
export function providerErrorEventFields(error: unknown): {
  error_kind?: ProviderErrorKind
  retryable?: boolean
  http_status?: number
  provider_code?: string
} {
  if (!(error instanceof ProviderError)) return {}
  return {
    error_kind: error.kind,
    retryable: error.retryable,
    ...(error.httpStatus !== undefined ? { http_status: error.httpStatus } : {}),
    ...(error.providerCode !== undefined ? { provider_code: error.providerCode } : {}),
  }
}

/** Provider errors expose their stable message without serializing the retained SDK cause. */
export function providerErrorMessage(error: unknown, fallback: (value: unknown) => string): string {
  return error instanceof ProviderError ? error.message : fallback(error)
}

File: rust/src/providers/provider_error.rs
use serde::{Deserialize, Serialize};

/// Stable provider-failure classification carried across the host/kernel boundary.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProviderErrorKind {
    Transport,
    Auth,
    RateLimit,
    ContextOverflow,
    InvalidRequest,
    Modality,
    ModelUnavailable,
    Protocol,
    Unknown,
}

impl ProviderErrorKind {
    pub(crate) const fn as_str(self) -> &'static str {
        match self {
            Self::Transport => "transport",
            Self::Auth => "auth",
            Self::RateLimit => "rate_limit",
            Self::ContextOverflow => "context_overflow",
            Self::InvalidRequest => "invalid_request",
            Self::Modality => "modality",
            Self::ModelUnavailable => "model_unavailable",
            Self::Protocol => "protocol",
            Self::Unknown => "unknown",
        }
    }
}

/// Structured provider error. Only these scalar fields cross into the canonical host event.
#[derive(Debug, Clone, thiserror::Error)]
#[error("{message}")]
pub struct ProviderError {
    pub provider: String,
    pub kind: ProviderErrorKind,
    pub retryable: bool,
    pub message: String,
    pub http_status: Option<u16>,
    pub provider_code: Option<String>,
}

impl ProviderError {
    pub fn new(
        provider: impl Into<String>,
        kind: ProviderErrorKind,
        retryable: bool,
        message: impl Into<String>,
    ) -> Self {
        Self {
            provider: provider.into(),
            kind,
            retryable,
            message: message.into(),
            http_status: None,
            provider_code: None,
        }
    }

    pub fn transport(provider: impl Into<String>, message: impl Into<String>) -> Self {
        Self::new(provider, ProviderErrorKind::Transport, true, message)
    }

    pub fn from_http(provider: impl Into<String>, status: u16, body: impl Into<String>) -> Self {
        let provider = provider.into();
        let body = body.into();
        let provider_code = provider_code(&body);
        let kind = classify_http(status, provider_code.as_deref());
        let retryable = matches!(
            kind,
            ProviderErrorKind::Transport
                | ProviderErrorKind::RateLimit
                | ProviderErrorKind::ModelUnavailable
        );
        Self {
            message: format!("{provider} HTTP {status}: {body}"),
            provider,
            kind,
            retryable,
            http_status: Some(status),
            provider_code,
        }
    }
}

fn provider_code(body: &str) -> Option<String> {
    let value: serde_json::Value = serde_json::from_str(body).ok()?;
    [
        value.pointer("/error/code"),
        value.pointer("/error/error_code"),
        value.pointer("/error/type"),
        value.get("code"),
        value.get("error_code"),
    ]
    .into_iter()
    .flatten()
    .find_map(|value| value.as_str().filter(|value| !value.is_empty()))
    .map(str::to_owned)
}

fn classify_http(status: u16, provider_code: Option<&str>) -> ProviderErrorKind {
    if status == 413
        || provider_code.is_some_and(|code| {
            code.eq_ignore_ascii_case("context_length_exceeded")
                || code.eq_ignore_ascii_case("prompt_too_long")
        })
    {
        return ProviderErrorKind::ContextOverflow;
    }
    match status {
        401 | 403 => ProviderErrorKind::Auth,
        429 => ProviderErrorKind::RateLimit,
        404 | 500..=599 => ProviderErrorKind::ModelUnavailable,
        408 | 409 => ProviderErrorKind::Transport,
        400 | 422 => ProviderErrorKind::InvalidRequest,
        _ => ProviderErrorKind::Unknown,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn http_context_overflow_uses_status_or_structured_code() {
        let by_status = ProviderError::from_http("test", 413, "too large");
        assert_eq!(by_status.kind, ProviderErrorKind::ContextOverflow);

        let by_code = ProviderError::from_http(
            "test",
            400,
            r#"{"error":{"code":"context_length_exceeded"}}"#,
        );
        assert_eq!(by_code.kind, ProviderErrorKind::ContextOverflow);
        assert_eq!(
            by_code.provider_code.as_deref(),
            Some("context_length_exceeded")
        );
    }

    #[test]
    fn prose_does_not_determine_failure_kind() {
        let error = ProviderError::from_http("test", 400, "413 prompt too long");
        assert_eq!(error.kind, ProviderErrorKind::InvalidRequest);
    }
}
